pkg/ruler/prom_client.go

```diff
@@ -0,0 +1,91 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log/level"
+	promapi "github.com/prometheus/client_golang/api"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+const orgIDHeader = "X-Scope-OrgId"
+
+type promClient struct {
+	promAPI promv1.API
+	timeout time.Duration
+}
+
+func NewPromClient(queryFrontendUrl, prometheusHTTPPrefix, orgID string, timeout time.Duration) *promClient {
+	if queryFrontendUrl == "" {
+		return nil
+	}
+	client, err := promapi.NewClient(promapi.Config{
+		Address:      queryFrontendUrl + prometheusHTTPPrefix,
+		RoundTripper: &addOrgIDRoundTripper{orgID: orgID, next: http.DefaultTransport},
+	})
+	if err != nil {
+		// use distributor queryable instead
```

Comment:
I didn't get it. Can we return error instead of nil?
Where do we use `distributor queryable` in the code

Reply:
My intention is if the _promclient_ is nil, we query to ingester using _engine.NewInstantQuery_.
Where do you think initialize the _promclient_ and how to address _promclient_ initialization error?